### DISCUSSION TITLE
Is it possible to run .NET in browser worklets?

### DISCUSSION THREAD

**Discussion by acaly** [dotnet/runtime#96024] [Q&A] [answered]
Worklets are loaded by the browser from a separate file. I am not familiar with blazor (especially the publishing process), but I think the runtime must first be loaded as a wasm module before my C# code can be executed. Is this supported or planned to be supported?

I found that someone achieved this with rust (speratus/web-synth-proto2), which uses javascript and `postMessage` to load wasm modules dynamically. Is this a good direction to try?

Thanks!

**Comment by maraf** [comment]
cc @pavelsavara

**Reply by pavelsavara** [reply]
Blazor is full UI framework and so the question is probably better aimed at dotnet runtime on WASM (Blazor WASM is running on it).

Running in worklet is not possible at the moment because we require dynamic `import()`.

I would like to hear more about why this useful idea.

We have similar topic for service workers here
https://github.com/dotnet/runtime/issues/95277

**Reply by acaly** [reply]
Thank you for the reply!

I agree that running blazor, which is an UI framework, inside a worklet is not of much interest. I asked because I was not aware of any other project type that outputs wasm in .NET.

I found that I can actually compile .NET to wasm with the experimental wasi-sdk. Is this what you meant by dotnet runtime on WASM? I believe once I have a standalone wasm module compiled from my C# code, I can do whatever the browser allows me to do in it.

**Reply by pavelsavara** [reply, marked as answer]
We have 2 flavors of dotnet for wasm, one is targeted to "OS" **Browser** and the other to **WASI**.

Here is simple dotnet on wasm demo I wrote for Net7. It will work for Net8 too. That is targeting browser JavaScript APIs.
https://github.com/pavelsavara/dotnet-wasm-todo-mvc

I think this is close to what you want. 
It would work in webworker (we are planning more specific demo soon https://github.com/dotnet/runtime/issues/95452)
But running it in worklet could be more challenging. Keep us posted 😉 

The WASI flavor is much less mature. Also the browser ecosystem is not very ready to run WASI. You can read more about our WASI thought here https://devblogs.microsoft.com/dotnet/extending-web-assembly-to-the-cloud/
